### PR TITLE
feat: added email sent signal to edx ace

### PR DESCRIPTION
**Pull request by AhtishamShahid**
## Description 

Added signal to track if email is sent for analytics and tracking.

## Ticket 
https://2u-internal.atlassian.net/browse/INF-1456

### CODE REVIEW

File: edx_ace/ace.py [hunk 1]
```diff
@@ -51,6 +51,7 @@ def send(msg):
         try:
             rendered_message = presentation.render(channel, msg)
             delivery.deliver(channel, rendered_message, msg)
+
```
Comment (OmarIthawi) [inline edx_ace/ace.py]: Please remove this line to keep the commit more focused.

File: edx_ace/delivery.py [hunk 2]
```diff
@@ -10,6 +10,7 @@
 from django.conf import settings
 
 from edx_ace.errors import RecoverableChannelDeliveryError
+from edx_ace.signals import ACE_EMAIL_SENT
```
Comment (OmarIthawi) [inline edx_ace/delivery.py]: @AhtishamShahid Should we use a Django signal here or a https://github.com/openedx/openedx-events and https://github.com/openedx/openedx-filters event?

This is more of discussion than a request. Please let me know what do you think?

File: edx_ace/signals.py [hunk 3]
```diff
@@ -0,0 +1,7 @@
+"""
+Signals for the edx_ace app
+"""
+from django.dispatch import Signal
+
+# signal to indicate that an email has been sent using ace
+ACE_EMAIL_SENT = Signal()
```
Comment (OmarIthawi) [inline edx_ace/signals.py]:
```suggestion
# signal to indicate that a message has been sent using ace
ACE_MESSAGE_SENT = Signal()
```

We support both Push notifications and email as message types among other types.

**Review by asadazam93 — APPROVED** [review]

File: edx_ace/signals.py [hunk 4]
```diff
@@ -0,0 +1,7 @@
+"""
+Signals for the edx_ace app
+"""
+from django.dispatch import Signal
+
+# signal to indicate that a message has been sent using ace
+ACE_EMAIL_SENT = Signal()
```
Comment (OmarIthawi) [inline edx_ace/signals.py]:
```suggestion
ACE_MESSAGE_SENT = Signal()
```

**Review by OmarIthawi — CHANGES_REQUESTED** [review]
Please rename it. I still see the old `EMAIL` word in the diff.

File: edx_ace/delivery.py (hunk 2 above)
Comment (AhtishamShahid) [reply edx_ace/delivery.py]: Hi,

We are not fundamentally changing the application behavior or flow , so using openedx-filters does not seem appropriate. However, openedx-events can work similarly to our current implementation. The use of global signals would be meaningful if we intended to handle that signal in other packages as well, but that is not the case at this time. Therefore, directly using Django signals makes more sense to me.

File: edx_ace/delivery.py [hunk 5]
```diff
@@ -60,8 +61,10 @@ def deliver(channel, rendered_message, message):
             message.report(f'{channel_type}_delivery_retried', num_seconds)
         else:
             message.report(f'{channel_type}_delivery_succeeded', True)
+            ACE_MESSAGE_SENT.send(sender=channel, message=message)
             return
 
     delivery_expired_report = f'{channel_type}_delivery_expired'
     logger.debug(delivery_expired_report)
     message.report(delivery_expired_report, get_current_time() - start_time)
+    ACE_MESSAGE_SENT.send(sender=channel, message=message)
```
Comment (OmarIthawi) [inline edx_ace/delivery.py:70]: Would you mind explaining the logic behind this line?

It seems to be added by mistake.

**Review by muhammadadeeltajamul — APPROVED** [review]
Have you thought of a way to check if an email fails to send?  I think they are also needed for tracking.

FYI:  I know the ticket scope is to generate event when email is send successfully.

**Review by OmarIthawi — COMMENTED** [review]
@AhtishamShahid I've sumitted a review already, but it seems that a bad internet connection got it lost.

I was surprised that it's been merged, but since my review was lost by GitHub, it's fine.

Would you mind checking the comment? It seems that we're sending the signal in cases where the message isn't sent.

Anyway, please ensure tests are added for the following cases:

 - Email is sent: ensure only one signal is sent.
 - Email sending fails: Ensure no signal is sent.
 - Email and Push notifications: Two signals are sent with appropriate `channel` parameter


```
ACE_MESSAGE_SENT.send(sender=channel, message=message)
```

File: edx_ace/delivery.py (hunk 5 above)
Comment (OmarIthawi) [reply edx_ace/delivery.py:70]: @muhammadadeeltajamul please check the review above and let me know when the notes can be addressed.

**Comment by AhtishamShahid** [comment]
> @AhtishamShahid I've sumitted a review already, but it seems that a bad internet connection got it lost.
> 
> I was surprised that it's been merged, but since my review was lost by GitHub, it's fine.
> 
> Would you mind checking the comment? It seems that we're sending the signal in cases where the message isn't sent.
> 
> Anyway, please ensure tests are added for the following cases:
> 
> * Email is sent: ensure only one signal is sent.
> * Email sending fails: Ensure no signal is sent.
> * Email and Push notifications: Two signals are sent with appropriate `channel` parameter
> 
> ```
> ACE_MESSAGE_SENT.send(sender=channel, message=message)
> ```

Create a PR with these changes https://github.com/openedx/edx-ace/pull/296

**Pull request merged**